File: src/Sum_Root_to_Leaf_Numbers.java
/**
 * 2014-6-13
 * @author Alvin
 * 
 */
import java.util.Vector;
public class Sum_Root_to_Leaf_Numbers {

	private Vector<Integer> vector = new Vector<Integer>();
	public int sumNumbers(TreeNode root) {
	     int result = 0 ;
	     if(root == null)
	    	 return result;
	    
	     String s ="";
	     saveNumber(s,root);
	     for(int i=0;i<vector.size();i++){
	    	 result += vector.get(i);
	     }
	     return result;
	}
	public void saveNumber(String s, TreeNode node){
		s += node.val;
		if(node.left == null && node.right ==null){
			vector.add(Integer.parseInt(s));
		}
		else{
			
			if(node.left != null)
				 saveNumber(s,node.left);
			if(node.right != null)
				 saveNumber(s,node.right);
		}
	}
	public static void main(String[] args) {
		TreeNode root = new TreeNode(1);
		TreeNode p2 = new TreeNode(2);
		TreeNode p3 = new TreeNode(3);
		root.left = p2;
		root.right = p3;
		System.out.println(new Sum_Root_to_Leaf_Numbers().sumNumbers(root));
		
	}

}

File: src/cpp/sum-root-to-leaf-numbers.cpp
#include <iostream>
#include <vector>
using namespace std;
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int sumNumbers(TreeNode *root) 
    {
        if ( root != NULL ) return innerSum( root , 0 );
        else return 0;    
    }
    int innerSum( TreeNode* node , int parent )
    {
        int valuea = 0 , valueb = 0;
        if( node->left == NULL && node->right == NULL ) 
            return parent*10 + node->val;

        if ( node->left != NULL ) valuea = innerSum(node->left, parent*10 + node->val);
        if ( node->right != NULL ) valueb = innerSum(node->right, parent*10 + node->val);
        //printf("%d,valuea=%d,valueb=%d,node->value=%d\n",parent,valuea , valueb,node->val );
        //valuea = valuea == 0 ? 0 : parent * 10 + valuea;
        //valueb = valueb == 0 ? 0 : parent * 10 + valueb;
        return  valuea + valueb;
    }
};
TreeNode* getParent(vector<TreeNode*> &v,int index)
{
	return v[(index - 1) >> 1];
}

TreeNode* arr2tree( int value[],int len )
{
	std::vector<TreeNode*> v;
	TreeNode* head = NULL;
	for (int i = 0; i < len; ++i)
	{
		TreeNode* node = new TreeNode(i+1);
		v.push_back(node);	
		if ( head == NULL ) head = node;
		else
		{
			TreeNode* parent = getParent(v,i);
			if ( i % 2 == 1 ) parent->left = node;
			else parent->right = node;
		}
	}
	return head;
}
void printVector2(vector<vector<int> > &v)
{
    char buffer[1024];
    memset( buffer,0,sizeof(char)*1024 );
    for (int i = 0; i < v.size(); ++i)
    {
        for (int j = 0; j < v[i].size(); ++j)
        {
            sprintf(buffer,"%s,%d",buffer,v[i][j]);
        }
        sprintf(buffer,"%s\n",buffer);
    }
    printf("%s\n",buffer );
}

int main(int argc, char const *argv[])
{
	int value[] = {1,2,3,4,5,6};
	TreeNode* root = arr2tree(value,6);
	Solution s;
	int result = s.sumNumbers(root);

	printf("result=%d\n",result);
	return 0;
}